File: backend/go/internal/services/account_abstraction_service.go
package services

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"fmt"
	"math/big"
	"time"

	"github.com/google/uuid"
	"github.com/tigerwallet/backend/internal/models"
)

type AccountAbstractionService struct {
	db *sql.DB
}

func NewAccountAbstractionService(db *sql.DB) *AccountAbstractionService {
	return &AccountAbstractionService{db: db}
}

// Create smart contract wallet (account abstraction)
func (s *AccountAbstractionService) CreateAccount(ctx context.Context, userID uuid.UUID, ownerAddress string, salt string) (*models.SmartAccount, error) {
	// Derive account address from owner and salt
	accountAddress := deriveSmartAccountAddress(ownerAddress, salt)

	account := &models.SmartAccount{
		ID:             uuid.New(),
		UserID:         userID,
		AccountAddress: accountAddress,
		OwnerAddress:   ownerAddress,
		Nonce:          0,
		Threshold:      1,
		Status:         "ACTIVE",
		Deployed:       false,
		CreatedAt:      time.Now(),
	}

	_, err := s.db.ExecContext(ctx, `
		INSERT INTO smart_accounts (id, user_id, account_address, owner_address, nonce, threshold, status, deployed, created_at)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
	`, account.ID, account.UserID, account.AccountAddress, account.OwnerAddress, 
		account.Nonce, account.Threshold, account.Status, account.Deployed, account.CreatedAt)

	if err != nil {
		return nil, fmt.Errorf("failed to create account: %w", err)
	}

	return account, nil
}

// Execute user operation (meta-transaction)
func (s *AccountAbstractionService) ExecuteUserOp(ctx context.Context, userOp *models.UserOperation) (*models.UserOperation, error) {
	// Verify signature
	if !s.verifySignature(userOp) {
		return nil, fmt.Errorf("invalid signature")
	}

	// Get account
	var account models.SmartAccount
	err := s.db.QueryRowContext(ctx, `
		SELECT id, nonce, deployed FROM smart_accounts 
		WHERE account_address = $1 AND user_id = $2
	`, userOp.Sender, userOp.UserID).Scan(&account.ID, &account.Nonce, &account.Deployed)

	if err == sql.ErrNoRows {
		return nil, fmt.Errorf("account not found")
	}
	if err != nil {
		return nil, err
	}

	// Check nonce
	if userOp.Nonce != account.Nonce {
		return nil, fmt.Errorf("invalid nonce")
	}

	// In production:
	// 1. Simulate the transaction
	// 2. Check paymaster sponsorship
	// 3. Execute via EntryPoint contract
	// 4. Update nonce

	// Update nonce
	_, err = s.db.ExecContext(ctx, `
		UPDATE smart_accounts SET nonce = nonce + 1 WHERE account_address = $1
	`, userOp.Sender)

	if err != nil {
		return nil, fmt.Errorf("failed to update nonce: %w", err)
	}

	// Mark userOp as confirmed
	userOp.Status = "CONFIRMED"
	userOp.ConfirmedAt = time.Now()

	return userOp, nil
}

// Add signer (multisig support)
func (s *AccountAbstractionService) AddSigner(ctx context.Context, userID uuid.UUID, accountAddress, signerAddress string, weight int) error {
	signerID := uuid.New()
	_, err := s.db.ExecContext(ctx, `
		INSERT INTO account_signers (id, user_id, account_address, signer_address, weight, status, added_at)
		VALUES ($1, $2, $3, $4, $5, $6, $7)
	`, signerID, userID, accountAddress, signerAddress, weight, "ACTIVE", time.Now())

	return err
}

// Remove signer
func (s *AccountAbstractionService) RemoveSigner(ctx context.Context, userID uuid.UUID, accountAddress, signerAddress string) error {
	_, err := s.db.ExecContext(ctx, `
		UPDATE account_signers SET status = 'REMOVED'
		WHERE account_address = $1 AND signer_address = $2 AND user_id = $3
	`, accountAddress, signerAddress, userID)

	return err
}

// Get account signers
func (s *AccountAbstractionService) GetSigners(ctx context.Context, accountAddress string) ([]models.AccountSigner, error) {
	rows, err := s.db.QueryContext(ctx, `
		SELECT id, signer_address, weight, status FROM account_signers 
		WHERE account_address = $1 AND status = 'ACTIVE'
	`, accountAddress)

	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var signers []models.AccountSigner
	for rows.Next() {
		var signer models.AccountSigner
		err := rows.Scan(&signer.ID, &signer.SignerAddress, &signer.Weight, &signer.Status)
		if err != nil {
			continue
		}
		signers = append(signers, signer)
	}

	return signers, nil
}

// Set up session key (for recurring transactions)
func (s *AccountAbstractionService) AddSessionKey(ctx context.Context, userID uuid.UUID, accountAddress, sessionKey string, permissions string, expiresAt time.Time) error {
	sessionID := uuid.New()
	_, err := s.db.ExecContext(ctx, `
		INSERT INTO session_keys (id, user_id, account_address, session_key, permissions, expires_at, created_at)
		VALUES ($1, $2, $3, $4, $5, $6, NOW())
	`, sessionID, userID, accountAddress, sessionKey, permissions, expiresAt)

	return err
}

// Verify session key
func (s *AccountAbstractionService) VerifySessionKey(ctx context.Context, accountAddress, sessionKey string) (bool, error) {
	var count int
	err := s.db.QueryRowContext(ctx, `
		SELECT COUNT(*) FROM session_keys 
		WHERE account_address = $1 AND session_key = $2 AND expires_at > NOW()
	`, accountAddress, sessionKey).Scan(&count)

	return count > 0, err
}

// Get user's smart accounts
func (s *AccountAbstractionService) GetUserAccounts(ctx context.Context, userID uuid.UUID) ([]models.SmartAccount, error) {
	rows, err := s.db.QueryContext(ctx, `
		SELECT id, account_address, owner_address, nonce, threshold, status, deployed, created_at
		FROM smart_accounts WHERE user_id = $1
	`, userID)

	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var accounts []models.SmartAccount
	for rows.Next() {
		var a models.SmartAccount
		err := rows.Scan(&a.ID, &a.AccountAddress, &a.OwnerAddress, 
			&a.Nonce, &a.Threshold, &a.Status, &a.Deployed, &a.CreatedAt)
		if err != nil {
			continue
		}
		accounts = append(accounts, a)
	}

	return accounts, nil
}

// Get user operations history
func (s *AccountAbstractionService) GetUserOps(ctx context.Context, accountAddress string) ([]models.UserOperation, error) {
	rows, err := s.db.QueryContext(ctx, `
		SELECT id, user_op_hash, sender, nonce, init_code, call_data, call_gas_limit, verification_gas_limit, pre_verification_gas, max_fee_per_gas, max_priority_fee_per_gas, signature, status, created_at
		FROM user_operations WHERE sender = $1 ORDER BY created_at DESC LIMIT 50
	`, accountAddress)

	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var ops []models.UserOperation
	for rows.Next() {
		var op models.UserOperation
		err := rows.Scan(&op.ID, &op.UserOpHash, &op.Sender, &op.Nonce, 
			&op.InitCode, &op.CallData, &op.CallGasLimit, &op.VerificationGasLimit,
			&op.PreVerificationGas, &op.MaxFeePerGas, &op.MaxPriorityFeePerGas,
			&op.Signature, &op.Status, &op.CreatedAt)
		if err != nil {
			continue
		}
		ops = append(ops, op)
	}

	return ops, nil
}

// Deploy account (bundle with first transaction)
func (s *AccountAbstractionService) DeployAccount(ctx context.Context, userID uuid.UUID, accountAddress string) error {
	_, err := s.db.ExecContext(ctx, `
		UPDATE smart_accounts SET deployed = true WHERE account_address = $1 AND user_id = $2
	`, accountAddress, userID)

	return err
}

// Enable paymaster sponsorship
func (s *AccountAbstractionService) EnablePaymaster(ctx context.Context, userID uuid.UUID, accountAddress, paymasterAddress string) error {
	_, err := s.db.ExecContext(ctx, `
		INSERT INTO paymaster_sponsors (id, user_id, account_address, paymaster_address, enabled_at)
		VALUES ($1, $2, $3, $4, NOW())
	`, uuid.New(), userID, accountAddress, paymasterAddress)

	return err
}

// Helper functions
func deriveSmartAccountAddress(owner, salt string) string {
	data := owner + salt
	hash := sha256.Sum256([]byte(data))
	return "0x" + hex.EncodeToString(hash[:20])
}

func (s *AccountAbstractionService) verifySignature(userOp *models.UserOperation) bool {
	// In production, verify EIP-4337 signature
	// Hash the user operation and verify against owner's signature
	data := fmt.Sprintf("%s%d%s%s", userOp.Sender, userOp.Nonce, userOp.CallData, userOp.Signature)
	hash := sha256.Sum256([]byte(data))
	
	// Simplified - in production use proper signature verification
	return len(userOp.Signature) > 0
}

func parseBigInt(s string) *big.Int {
	i, _ := new(big.Int).SetString(s, 0)
	return i
}

File: mobile/flutter/lib/features/account_abstraction/account_abstraction_service.dart
// Account Abstraction Service - Flutter Mobile

import 'dart:convert';
import 'package:http/http.dart' as http;

class AccountAbstractionService {
  static const String API_BASE = 'https://api.tigerwallet.com/api/v1';
  String? _token;
  
  AccountAbstractionService({String? token}) : _token = token;
  
  Map<String, String> get _headers => {
    'Content-Type': 'application/json',
    if (_token != null) 'Authorization': 'Bearer $_token',
  };
  
  // Create smart account
  Future<SmartAccount> createAccount({
    required String ownerAddress,
    String? salt,
  }) async {
    final response = await http.post(
      Uri.parse('$API_BASE/account/create'),
      headers: _headers,
      body: json.encode({
        'ownerAddress': ownerAddress,
        'salt': salt,
      }),
    );
    
    if (response.statusCode == 201) {
      final data = json.decode(response.body);
      return SmartAccount.fromJson(data['data']);
    }
    throw Exception('Failed to create account');
  }
  
  // Execute user operation
  Future<UserOperation> executeOp(UserOperation op) async {
    final response = await http.post(
      Uri.parse('$API_BASE/account/execute'),
      headers: _headers,
      body: json.encode(op.toJson()),
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return UserOperation.fromJson(data['data']);
    }
    throw Exception('Failed to execute operation');
  }
  
  // Get user's accounts
  Future<List<SmartAccount>> getAccounts() async {
    final response = await http.get(
      Uri.parse('$API_BASE/account/list'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return (data['data'] as List).map((a) => SmartAccount.fromJson(a)).toList();
    }
    return [];
  }
  
  // Get account details
  Future<SmartAccount> getAccount(String address) async {
    final response = await http.get(
      Uri.parse('$API_BASE/account/$address'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return SmartAccount.fromJson(data['data']);
    }
    throw Exception('Failed to get account');
  }
  
  // Add signer
  Future<bool> addSigner(String accountAddress, String signerAddress, int weight) async {
    final response = await http.post(
      Uri.parse('$API_BASE/account/signers'),
      headers: _headers,
      body: json.encode({
        'accountAddress': accountAddress,
        'signerAddress': signerAddress,
        'weight': weight,
      }),
    );
    
    return response.statusCode == 201;
  }
  
  // Remove signer
  Future<bool> removeSigner(String accountAddress, String signerAddress) async {
    final response = await http.delete(
      Uri.parse('$API_BASE/account/signers?account=$accountAddress&signer=$signerAddress'),
      headers: _headers,
    );
    
    return response.statusCode == 200;
  }
  
  // Get signers
  Future<List<AccountSigner>> getSigners(String accountAddress) async {
    final response = await http.get(
      Uri.parse('$API_BASE/account/signers?account=$accountAddress'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return (data['data'] as List).map((s) => AccountSigner.fromJson(s)).toList();
    }
    return [];
  }
  
  // Add session key
  Future<bool> addSessionKey(String accountAddress, String sessionKey, Map<String, dynamic> permissions, DateTime expiresAt) async {
    final response = await http.post(
      Uri.parse('$API_BASE/account/sessions'),
      headers: _headers,
      body: json.encode({
        'accountAddress': accountAddress,
        'sessionKey': sessionKey,
        'permissions': permissions,
        'expiresAt': expiresAt.toIso8601String(),
      }),
    );
    
    return response.statusCode == 201;
  }
  
  // Verify session key
  Future<bool> verifySessionKey(String accountAddress, String sessionKey) async {
    final response = await http.get(
      Uri.parse('$API_BASE/account/sessions/verify?account=$accountAddress&key=$sessionKey'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return data['data'] ?? false;
    }
    return false;
  }
  
  // Get user operations
  Future<List<UserOperation>> getUserOps(String accountAddress) async {
    final response = await http.get(
      Uri.parse('$API_BASE/account/ops?account=$accountAddress'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return (data['data'] as List).map((o) => UserOperation.fromJson(o)).toList();
    }
    return [];
  }
  
  // Enable paymaster
  Future<bool> enablePaymaster(String accountAddress, String paymasterAddress) async {
    final response = await http.post(
      Uri.parse('$API_BASE/account/paymaster'),
      headers: _headers,
      body: json.encode({
        'accountAddress': accountAddress,
        'paymasterAddress': paymasterAddress,
      }),
    );
    
    return response.statusCode == 201;
  }
}

class SmartAccount {
  final String id;
  final String accountAddress;
  final String ownerAddress;
  final int nonce;
  final int threshold;
  final String status;
  final bool deployed;
  final DateTime createdAt;
  
  SmartAccount({
    required this.id,
    required this.accountAddress,
    required this.ownerAddress,
    required this.nonce,
    required this.threshold,
    required this.status,
    required this.deployed,
    required this.createdAt,
  });
  
  factory SmartAccount.fromJson(Map<String, dynamic> json) {
    return SmartAccount(
      id: json['id'] ?? '',
      accountAddress: json['accountAddress'] ?? '',
      ownerAddress: json['ownerAddress'] ?? '',
      nonce: json['nonce'] ?? 0,
      threshold: json['threshold'] ?? 1,
      status: json['status'] ?? 'ACTIVE',
      deployed: json['deployed'] ?? false,
      createdAt: DateTime.parse(json['createdAt']),
    );
  }
}

class AccountSigner {
  final String id;
  final String signerAddress;
  final int weight;
  final String status;
  
  AccountSigner({
    required this.id,
    required this.signerAddress,
    required this.weight,
    required this.status,
  });
  
  factory AccountSigner.fromJson(Map<String, dynamic> json) {
    return AccountSigner(
      id: json['id'] ?? '',
      signerAddress: json['signerAddress'] ?? '',
      weight: json['weight'] ?? 1,
      status: json['status'] ?? 'ACTIVE',
    );
  }
}

class UserOperation {
  final String id;
  final String userOpHash;
  final String sender;
  final int nonce;
  final String? initCode;
  final String callData;
  final int callGasLimit;
  final int verificationGasLimit;
  final int preVerificationGas;
  final String maxFeePerGas;
  final String maxPriorityFeePerGas;
  final String signature;
  final String status;
  final DateTime createdAt;
  
  UserOperation({
    required this.id,
    required this.userOpHash,
    required this.sender,
    required this.nonce,
    this.initCode,
    required this.callData,
    required this.callGasLimit,
    required this.verificationGasLimit,
    required this.preVerificationGas,
    required this.maxFeePerGas,
    required this.maxPriorityFeePerGas,
    required this.signature,
    required this.status,
    required this.createdAt,
  });
  
  factory UserOperation.fromJson(Map<String, dynamic> json) {
    return UserOperation(
      id: json['id'] ?? '',
      userOpHash: json['userOpHash'] ?? '',
      sender: json['sender'] ?? '',
      nonce: json['nonce'] ?? 0,
      initCode: json['initCode'],
      callData: json['callData'] ?? '',
      callGasLimit: json['callGasLimit'] ?? 21000,
      verificationGasLimit: json['verificationGasLimit'] ?? 100000,
      preVerificationGas: json['preVerificationGas'] ?? 21000,
      maxFeePerGas: json['maxFeePerGas'] ?? '0',
      maxPriorityFeePerGas: json['maxPriorityFeePerGas'] ?? '0',
      signature: json['signature'] ?? '',
      status: json['status'] ?? 'PENDING',
      createdAt: DateTime.parse(json['createdAt']),
    );
  }
  
  Map<String, dynamic> toJson() => {
    'sender': sender,
    'nonce': nonce,
    'initCode': initCode,
    'callData': callData,
    'callGasLimit': callGasLimit,
    'verificationGasLimit': verificationGasLimit,
    'preVerificationGas': preVerificationGas,
    'maxFeePerGas': maxFeePerGas,
    'maxPriorityFeePerGas': maxPriorityFeePerGas,
    'signature': signature,
  };
}
